File: examples/kerberos.rs
use std::error::Error;

use base64::Engine;
use proptest::char::any;
use reqwest::header::{
    ACCEPT, ACCEPT_ENCODING, ACCEPT_LANGUAGE, AUTHORIZATION, CONNECTION, CONTENT_LENGTH, HOST, USER_AGENT,
    WWW_AUTHENTICATE,
};
use reqwest::StatusCode;
use sspi::{
    AcquireCredentialsHandleResult, BufferType, ClientRequestFlags, CredentialsBuffers, DataRepresentation,
    EncryptionFlags, InitializeSecurityContextResult, Kerberos, KerberosConfig, SecurityBuffer, SecurityBufferRef,
    SecurityStatus, Sspi, SspiImpl, Username,
};
use tracing_subscriber::prelude::*;
use tracing_subscriber::{fmt, EnvFilter};

fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let kdc_url = std::env::var("SSPI_KDC_URL").expect("missing KDC URL set in SSPI_KDC_URL"); //tcp://ad-compter-name.domain:88
    let hostname = std::env::var("SSPI_WINRM_HOST").expect("missing host name set in SSPI_WINRM_HOST"); // winrm_server_name.domain
    let username = std::env::var("SSPI_WINRM_USER").expect("missing username set in SSPI_WINRM_USER"); // username@domain
    let password = std::env::var("SSPI_WINRM_PASS").expect("missing password set in SSPI_WINRM_PASS");
    let auth_method = std::env::var("SSPI_WINRM_AUTH").expect("missing auth METHOD set in SSPI_WINRM_AUTH"); // Negotiate or Kerberos

    tracing_subscriber::registry()
        .with(fmt::layer())
        .with(EnvFilter::from_env("SSPI_LOG_LEVEL"))
        .init();

    let kerberos_config = KerberosConfig::new(&kdc_url, hostname.clone());
    let mut kerberos = Kerberos::new_client_from_config(kerberos_config).unwrap();

    let mut acq_creds_handle_result = get_cred_handle(&mut kerberos, username, password);

    let mut input_token = String::new();
    let mut client = reqwest::blocking::Client::new(); // super IMPORTANT, KEEP-ALIVE the http connection!
    let output_token = loop {
        tracing::info!(?input_token, "step input token");
        let (output_token, status) = step(
            &mut kerberos,
            &mut acq_creds_handle_result.credentials_handle,
            &input_token,
            &hostname,
        );
        tracing::info!(?output_token, ?status, "step result");
        if status == SecurityStatus::ContinueNeeded || status == SecurityStatus::Ok {
            let (token_from_server, status_code) =
                process_authentication(&output_token, &mut client, &auth_method, &hostname)?;

            if status == SecurityStatus::Ok {
                tracing::info!(?token_from_server, "Authentication completed successfully");
                break output_token;
            }

            input_token = token_from_server;
        } else {
            panic!("Having problem continue authentication");
        }
    };

    tracing::info!(?output_token, "Authentication completed successfully");

    let mut request = std::fs::read("./soap.xml")?;

    let mut token = vec![0u8; kerberos.query_context_sizes()?.security_trailer as usize];
    let mut buffers = vec![
        SecurityBufferRef::data_buf(&mut request),
        SecurityBufferRef::token_buf(&mut token),
    ];

    let sec_status = kerberos.encrypt_message(EncryptionFlags::empty(), &mut buffers, 0)?;

    tracing::info!(?sec_status, "Encrypting message");

    // send_http(&hostname, &mut client, None, Some());

    Ok(())
}

pub(crate) fn get_cred_handle(
    kerberos: &mut Kerberos,
    username: String,
    password: String,
) -> AcquireCredentialsHandleResult<Option<CredentialsBuffers>> {
    let identity = sspi::AuthIdentity {
        username: Username::parse(&username).expect("username is not in the correct format"),
        password: password.into(),
    };
    let acq_creds_handle_result = kerberos
        .acquire_credentials_handle()
        .with_credential_use(sspi::CredentialUse::Outbound)
        .with_auth_data(&identity.into())
        .execute(kerberos)
        .expect("AcquireCredentialsHandle resulted in error");
    acq_creds_handle_result
}

pub(crate) fn process_authentication(
    token_neeeds_to_be_sent: &String,
    client: &mut reqwest::blocking::Client,
    auth_method: &str,
    hostname: &str,
) -> Result<(String, StatusCode), Box<dyn std::error::Error + Send + Sync>> {
    let auth = format!("{} {}", auth_method, token_neeeds_to_be_sent);
    let server_result = send_http(hostname, client, Some(auth), None)?;

    let www_authenticate = server_result
        .headers()
        .get(WWW_AUTHENTICATE)
        .ok_or("expecting www-authentication header from server but not found")?;

    tracing::info!(?www_authenticate, "WWW-Authenticate header from server");

    let server_token = www_authenticate
        .to_str()
        .unwrap()
        .split_once(" ")
        .ok_or("expecting www-authentication header from server but not found")
        .unwrap()
        .1
        .to_owned();
    tracing::info!(?server_token, "Server token");

    Ok((server_token, server_result.status()))
}

pub(crate) fn send_http(
    hostname: &str,
    client: &mut reqwest::blocking::Client,
    authorization: Option<String>,
    body: Option<String>,
) -> Result<reqwest::blocking::Response, Box<dyn Error + Send + Sync>> {
    let builder = client
        .post(format!("http://{}:5985/wsman?PSVersion=7.3.8", hostname))
        .header(HOST, format!("{}:5985", hostname))
        .header(CONNECTION, "keep-alive")
        .header(CONTENT_LENGTH, "0")
        .header(
            USER_AGENT,
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/118.0",
        )
        .header(ACCEPT, "*/*")
        .header(ACCEPT_ENCODING, "gzip, deflate")
        .header(ACCEPT_LANGUAGE, "en-US,en;q=0.9");

    let builder = if let Some(auth) = authorization {
        builder.header(AUTHORIZATION, auth)
    } else {
        builder
    };

    let resp = if let Some(body) = body {
        builder.header(CONTENT_LENGTH, body.len().to_string()).body(body)
    } else {
        builder
    }
    .send()?;

    Ok(resp)
}

fn step_helper(
    kerberos: &mut Kerberos,
    cred_handle: &mut <Kerberos as SspiImpl>::CredentialsHandle,
    input_buffer: &mut [SecurityBuffer],
    output_buffer: &mut [SecurityBuffer],
    hostname: &str,
) -> Result<InitializeSecurityContextResult, Box<dyn std::error::Error>> {
    let target_name = format!("HTTP/{}", hostname);
    let mut builder = kerberos
        .initialize_security_context()
        .with_credentials_handle(cred_handle)
        .with_context_requirements(ClientRequestFlags::INTEGRITY | ClientRequestFlags::CONFIDENTIALITY)
        .with_target_data_representation(DataRepresentation::Native)
        .with_target_name(&target_name)
        .with_input(input_buffer)
        .with_output(output_buffer);

    let result = kerberos
        .initialize_security_context_impl(&mut builder)?
        .resolve_with_default_network_client()?;

    Ok(result)
}

pub fn step(
    kerberos: &mut Kerberos,
    cred_handle: &mut <Kerberos as SspiImpl>::CredentialsHandle,
    input_token: &String,
    hostname: &str,
) -> (String, SecurityStatus) {
    let input_buffer = base64::engine::general_purpose::STANDARD.decode(input_token).unwrap();
    let mut secure_input_buffer = vec![SecurityBuffer::new(input_buffer, BufferType::Token)];
    let mut secure_output_buffer = vec![SecurityBuffer::new(Vec::new(), BufferType::Token)];
    match step_helper(
        kerberos,
        cred_handle,
        &mut secure_input_buffer,
        &mut secure_output_buffer,
        hostname,
    ) {
        Ok(result) => {
            let output_buffer = secure_output_buffer[0].to_owned();
            (
                base64::engine::general_purpose::STANDARD.encode(output_buffer.buffer),
                result.status,
            )
        }
        Err(e) => {
            panic!("Error in step: {}", e);
        }
    }
}

File: examples/kerberos/kerberos.py
# filepath: c:\dev\sspi-rs\examples\kerberos\kerberos.py
import base64
import os
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple, Union, ByteString

import requests
import sspicon
import win32security


def get_credentials(username: str, domain: str, password: str):
    """
    Acquire Kerberos credentials using the provided username, domain, and password.

    Returns:
        Tuple containing the credential handle and expiration time.
    """
    return win32security.AcquireCredentialsHandle(
        None,
        "Kerberos",
        win32security.SECPKG_CRED_OUTBOUND,
        None,
        (username, domain, password),
    )


def send_http(
    hostname: str,
    session: requests.Session,
    auth: Optional[str] = None,
    body: Optional[Union[str, bytes]] = None,
) -> requests.Response:
    """
    Send an HTTP request to the specified host.

    Args:
        hostname: Target hostname
        session: Requests session object
        auth: Optional authorization header value
        body: Optional request body

    Returns:
        HTTP response object
    """
    url = f"http://{hostname}:5985/wsman?PSVersion=7.3.8"
    headers: Dict[str, str] = {
        "HOST": f"{hostname}:5985",
        "CONNECTION": "keep-alive",
        "CONTENT-LENGTH": str(len(body) if body else 0),
        "USER-AGENT": "Mozilla/5.0",
        "ACCEPT": "*/*",
        "ACCEPT-ENCODING": "gzip, deflate",
        "ACCEPT-LANGUAGE": "en-US,en;q=0.9",
    }
    if auth:
        headers["AUTHORIZATION"] = auth

    resp = session.post(url, headers=headers, data=body or "")
    return resp


def process_authentication(
    token_to_send: str, session: requests.Session, auth_method: str, hostname: str
) -> Tuple[str, int]:
    """
    Process authentication by sending a token and getting a response.

    Args:
        token_to_send: Authentication token to send
        session: Requests session
        auth_method: Authentication method (e.g., "Negotiate")
        hostname: Target hostname

    Returns:
        Tuple containing the server token and status code
    """
    auth_header = f"{auth_method} {token_to_send}"
    response = send_http(hostname, session, auth_header)

    www_auth = response.headers.get("WWW-Authenticate")
    if not www_auth or "Negotiate" not in www_auth:
        raise Exception("expecting www-authentication header from server but not found")

    parts = www_auth.split(" ", 1)
    if len(parts) != 2:
        raise Exception("Invalid WWW-Authenticate header format")

    server_token = parts[1].strip()
    return server_token, response.status_code


def step_sspi(
    cred_handle, target_spn: str, token_input: Optional[str] = None
) -> Tuple[str, int]:
    """
    Perform a step in the SSPI authentication process.

    Args:
        ctx_handle: Credential handle
        target_spn: Service Principal Name
        token_input: Optional input token

    Returns:
        Tuple containing the output token and status code
    """
    # Prepare input buffer
    if token_input:
        token_input_bytes: bytes = base64.b64decode(token_input)
        in_buf = win32security.PySecBufferDescType()
        token_buf = win32security.PySecBufferType(
            len(token_input_bytes), sspicon.SECBUFFER_TOKEN
        )
        token_buf.Buffer = token_input_bytes
        in_buf.append(token_buf)
    else:
        in_buf = None

    # Output buffer
    out_buf = win32security.PySecBufferDescType()
    out_token_buf = win32security.PySecBufferType(0, sspicon.SECBUFFER_TOKEN)
    out_buf.append(out_token_buf)
    context = getattr(step_sspi, "context", None)  # Initialize security context
    result = win32security.InitializeSecurityContext(
        cred_handle,  # Credential
        context,  # Context
        target_spn,  # TargetName
        sspicon.ISC_REQ_CONFIDENTIALITY | sspicon.ISC_REQ_INTEGRITY,  # ContextReq
        sspicon.SECURITY_NATIVE_DREP,  # TargetDataRep
        in_buf,  # pInput
        None,  
        out_buf,  # pOutput
    )

    new_ctx, out_bufs, status, _ = result
    step_sspi.context = new_ctx

    for buf in out_bufs:
        if buf.type == win32security.SECBUFFER_TOKEN:
            return base64.b64encode(buf.Buffer).decode(), status

    return "", status


def main() -> None:
    """
    Main function that handles Kerberos authentication and encryption.
    """
    # kdc_url: str = os.environ["SSPI_KDC_URL"]  # Used for configuration
    hostname: str = os.environ["SSPI_WINRM_HOST"]
    username: str = os.environ["SSPI_WINRM_USER"]
    password: str = os.environ["SSPI_WINRM_PASS"]
    auth_method: str = os.environ["SSPI_WINRM_AUTH"]

    domain: str = username.split("@")[1] if "@" in username else ""
    user: str = username.split("@")[0]

    (cred_handle, time_type) = get_credentials(user, domain, password)
    session: requests.Session = requests.Session()
    input_token: str = ""

    while True:
        print(f"step input token: {input_token}")
        output_token: str
        status: int
        output_token, status = step_sspi(
            cred_handle=cred_handle,
            target_spn=f"HTTP/{hostname}",
            token_input=input_token,
        )
        print(f"step result: {output_token[:30]}..., status: {status}")

        if status in [0, 0x90312]:  # SEC_E_OK or SEC_I_CONTINUE_NEEDED
            server_token: str
            status_code: int
            server_token, status_code = process_authentication(
                output_token, session, auth_method, hostname
            )

            if status == 0:  # SEC_E_OK
                print(f"Authentication completed successfully: {server_token[:30]}...")
                break
            input_token = server_token
        else:
            raise RuntimeError("Problem continuing authentication")

    # Encryption phase (simulating encrypting ./soap.xml)
    if not hasattr(step_sspi, "ctx_data"):
        raise RuntimeError("Security context missing for encryption")

    xml_data: bytes = Path("./soap.xml").read_bytes()
    trailer_size: int = win32security.QueryContextAttributes(
        step_sspi.ctx_data, sspicon.SECPKG_ATTR_SIZES
    )["SecurityTrailer"]
    token: bytearray = bytearray(trailer_size)

    sec_buffers = win32security.PySecBufferDescType()
    sec_buffers.append(
        win32security.PySecBufferType(len(xml_data), win32security.SECBUFFER_DATA)
    )
    sec_buffers[-1].Buffer = xml_data
    sec_buffers.append(
        win32security.PySecBufferType(len(token), win32security.SECBUFFER_TOKEN)
    )

    win32security.EncryptMessage(step_sspi.ctx_data, 0, sec_buffers, 0)
    print("Encrypting message complete")


if __name__ == "__main__":
    main()
